File: go/sum_of_root_to_leaf_binary_numbers.go
package leetcode

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func sumRootToLeaf(root *TreeNode) int {
	return dfs(root, 0)
}

func dfs(node *TreeNode, curr int) int {
	if node == nil {
		return 0
	}

	curr <<= 1
	curr |= node.Val

	if node.Left == nil && node.Right == nil {
		return curr
	}
	return dfs(node.Left, curr) + dfs(node.Right, curr)
}

File: python/sum-of-root-to-leaf-binary-numbers.py
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        def dfs(node, curr):
            if not node: return 0
            
            curr <<= 1
            curr |= node.val
            
            if node.left == None and node.right == None:
                return curr
            
            return dfs(node.left, curr) + dfs(node.right, curr)
        
        return dfs(root, 0)